### PR TITLE
Adds setter for anchored. Cleans up a lot of anchored, density, and update_icon implementations

### PR DESCRIPTION
**Pull request by Tsar-Salat**
Ports:
- https://github.com/tgstation/tgstation/pull/52254

## About The Pull Request
Basically converts a bunch of `anchored = T/F` to `set_anchored(T/F)`, which is a direct upgrade.

Cleans up a shitton of anchored implementations. I also cleaned up a few density and icon calls because they were adjacent.

The largest cleanup occured in how library shelves anchored. It sucks a lot less

## Why It's Good For The Game
I looked at literally every implementation of `anchored =`

This is going to solve a lot of issues with this stupid var.

> Basically, creating a setter proc as opposed to just setting value on the var directly has the benefit of allowing us to append behavior whenever we call it.
> In this case and the prior https://github.com/BeeStation/BeeStation-Hornet/pull/8689, we are attaching a signal so it fires whenever the anchored value successfully changes.
> 
> With set_anchored now defined at the atom-level, we can move behavior that was before on snowflake procs, onto it.
> 
> For example, machines that are dependent on being anchored to function (Satellites, Pacman, etc.), this is useful, mostly for code cleanliness.
> 
> For the fix part of this pr, its because of how poorly bookcases handle states, akin to the mess that was https://github.com/BeeStation/BeeStation-Hornet/pull/9989

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

https://github.com/BeeStation/BeeStation-Hornet/assets/62388554/08df62f4-37ca-4182-b7fb-6778bb007410

</details>

## Changelog
:cl: RKz, ShizCalev
refactor: Converted everything to use setAnchored() as opposed to setting the anchored var directly. This means that COMSIG_MOVABLE_SETANCHORED can now actually be used reliably to track if an object's anchored var has changed.
fix: The refactor also means that varediting the anchored var will now properly update most items' other states that rely on it being set a certain way.
fix: Fixed bookcases having the incorrect icon when they're unanchored.
fix: Fixed bookcases not automatically picking up spellbooks / storage books when made/spawned.
fix: Fixed bookcases not dropping spellbooks / storage books when deconstructed.
/:cl:

### CODE REVIEW

File: code/__DEFINES/is_helpers.dm [hunk 1]
```diff
@@ -293,15 +293,22 @@ GLOBAL_LIST_INIT(glass_sheet_types, typecacheof(list(
...
 #define isshuttleturf(T) (length(T.baseturfs) && (/turf/baseturf_skipover/shuttle in T.baseturfs))
 
+#define isbook(O) (is_type_in_typecache(O, GLOB.book_types))
+
+GLOBAL_LIST_INIT(book_types, typecacheof(list(
+	/obj/item/book,
+	/obj/item/spellbook,
+	/obj/item/storage/book)))
+
+#define IS_EDIBLE(O) (O.GetComponent(/datum/component/edible))
```
Comment (itsmeow) [inline code/__DEFINES/is_helpers.dm]: duplicate def

File: code/modules/antagonists/cult/cult_structures.dm [hunk 2]
```diff
@@ -54,6 +54,25 @@
 	else
 		..()
 
+/*
+/obj/structure/destructible/cult/set_anchored(anchorvalue)
+	. = ..()
+	if(isnull(.))
+		return
+	update_icon()
```
Comment (itsmeow) [inline code/modules/antagonists/cult/cult_structures.dm]: what is this

File: code/modules/awaymissions/capture_the_flag.dm [hunk 3]
```diff
@@ -61,39 +61,41 @@
...
 	STOP_PROCESSING(SSobj, src)
-	..()
+	anchored = FALSE // Hacky usage that bypasses set_anchored(), because normal checks need this to be FALSE to pass
+	. = ..() //this is the actual normal item checks
+	if(.) //only apply these flag passives
+		anchored = TRUE // Avoid directly assigning to anchored and prefer to use set_anchored() on normal circumstances.
+		return
+	//passing means the user picked up the flag so we can now apply this
+	user.set_anchored(TRUE)
+	user.status_flags &= ~CANPUSH
```
Comment (itsmeow) [inline code/modules/awaymissions/capture_the_flag.dm:91]: is this tested?

File: code/modules/library/lib_items.dm [hunk 4]
```diff
@@ -30,71 +34,83 @@
...
 				to_chat(user, "<span class='notice'>You unwrench the frame.</span>")
-				anchored = FALSE
-				state = 0
+				set_anchored(FALSE)
 
-		if(2)
+		if(BOOKCASE_FINISHED)
 			var/datum/component/storage/STR = I.GetComponent(/datum/component/storage)
-			if(is_type_in_list(I, allowed_books))
+			if(isbook(I))
```
Comment (itsmeow) [inline code/modules/library/lib_items.dm]: behavior change

File: code/modules/library/lib_items.dm [hunk 5]
```diff
@@ -125,34 +141,42 @@
...
 /obj/structure/bookcase/deconstruct(disassembled = TRUE)
-	new /obj/item/stack/sheet/wood(loc, 4)
-	for(var/obj/item/book/B in contents)
-		B.forceMove(get_turf(src))
-	qdel(src)
-
-
-/obj/structure/bookcase/update_icon()
+	var/atom/Tsec = drop_location()
+	new /obj/item/stack/sheet/wood/(Tsec, 4)
```
Comment (itsmeow) [inline code/modules/library/lib_items.dm]:
```suggestion
	new /obj/item/stack/sheet/wood(Tsec, 4)
```

**Review by itsmeow — DISMISSED** [review]

File: code/modules/library/lib_items.dm (hunk 4 above)
Comment (Tsar-Salat) [reply code/modules/library/lib_items.dm]: This is why I hate old merge conflicts

File: code/__DEFINES/is_helpers.dm (hunk 1 above)
Comment (Tsar-Salat) [reply code/__DEFINES/is_helpers.dm]: Im the one who moved it originally too, man thats ironic.

removed in 3814797

File: code/modules/antagonists/cult/cult_structures.dm (hunk 2 above)
Comment (Tsar-Salat) [reply code/modules/antagonists/cult/cult_structures.dm]: something me of 3 months ago forgot to test

removed in 3814797

File: code/modules/library/lib_items.dm (hunk 4 above)
Comment (Tsar-Salat) [reply code/modules/library/lib_items.dm]: resolved in 3814797

File: code/modules/library/lib_items.dm (hunk 5 above)
Comment (Tsar-Salat) [reply code/modules/library/lib_items.dm]: fixed in 3814797

**Comment by Tsar-Salat** [comment]
im taking a break, Semester begins soon and this is affecting my mental health

File: code/modules/awaymissions/capture_the_flag.dm (hunk 3 above)
Comment (Tsar-Salat) [reply code/modules/awaymissions/capture_the_flag.dm:91]: Atomized into #10049

**Comment by itsmeow** [comment]
Is this still a draft?

**Comment by Tsar-Salat** [comment]
> Is this still a draft?

Its ready

**Review by Rukofamicom — CHANGES_REQUESTED** [review]
This PR needs a better explanation of how it improves the game than:

> This is going to solve a lot of issues with this stupid var.

Try to list the "lot of issues" this solves. 
The parent PR from TG does no better job of explaining this.

**Comment by Tsar-Salat** [comment]
> Try to list the "lot of issues" this solves.

I thought I was thorough in the changelog. If you'd like to me elaborate specifically on how this works, sure.

Basically, creating a setter proc as opposed to just setting value on the var directly has the benefit of allowing us to append behavior whenever we call it. 
In this case and the prior #8689, we are attaching a signal so it fires whenever the anchored value successfully changes. 

With set_anchored now defined at the atom-level, we can move behavior that was before on snowflake procs, onto it.

For example, machines that are dependent on being anchored to function (Satellites, Pacman, etc.), this is useful, mostly for code cleanliness.

**Comment by Tsar-Salat** [comment]
I made this PR in April, I dont know why I didnt show testing evidence for satellites and booksshelves. Probably because of the long draft period.

Ill do that when I have free time after my exams

**Review by Rukofamicom — APPROVED** [review]
I don't believe additional testing evidence is necessary in this case, and thank you for the detailed explanation.

**Pull request merged**